### PR TITLE
Save a ParGridFunction in Serial

### PR DESCRIPTION
**Pull request by kmittal2**
Follow up to #3035. This PR adds functionality to save a ParGridFunction as a serial gridfunction which is compatible with a serial mesh output using `ParMesh::PrintAsSerial`.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3588](https://github.com/mfem/mfem/pull/3588) | @kmittal2 | @v-dobrev | @atallah727 + @cjvogl | 5/23/23 | 6/21/23 | 6/26/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by v-dobrev** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by atallah727 — APPROVED** [review]

File: fem/pgridfunc.cpp [hunk 1]
```diff
@@ -908,6 +908,100 @@ void ParGridFunction::SaveAsOne(const char *fname, int precision) const
...
+
+GridFunction ParGridFunction::GetSerialGridFunction(int save_rank) const
+{
+   ParFiniteElementSpace *pfespace = ParFESpace();
+   ParMesh *pmesh = pfespace->GetParMesh();
+
+   Mesh serialmesh = pmesh->GetSerialMesh(save_rank);
+
+   int vdim = pfespace->GetVDim();
+   auto *fespace_serial = new FiniteElementSpace(&serialmesh,
```
Comment (cjvogl) [inline fem/pgridfunc.cpp:931]: Does this cause a memory leak?  If so, I'd ask for either:
- addition of `delete fespace_serial;` somewhere
- use of smart pointer, i.e., `std::unique_ptr<FiniteElementSpace> fespace_serial(new FiniteElementSpace(...));`

File: fem/pgridfunc.cpp [hunk 2]
```diff
@@ -908,6 +908,100 @@ void ParGridFunction::SaveAsOne(const char *fname, int precision) const
...
+
+   GridFunction gf_serial(fespace_serial);
+   Array<double> vals;
+   Array<int> dofs;
+   MPI_Status status;
+   int n_send_recv;
+
+   int MyRank = pmesh->GetMyRank(),
+       NRanks = pmesh->GetNRanks();
+   MPI_Comm MyComm = pmesh->GetComm();
```
Comment (cjvogl) [inline fem/pgridfunc.cpp:944]: There's a mixture of camel case, snake case, and pascal case throughout this PR... maybe worth making it more consistent but not a big deal.

**Comment by cjvogl** [comment]
@kmittal2, I added some inline comments and have some additional requests/comments:
- would it be possible to suggest a way to smoke test this new functionality?  I guess ideal such a test would be added to the MFEM unit tests.
- it looks like there are now merge conflicts with the master branch... perhaps these should be resolved before testing?

File: fem/pgridfunc.cpp (hunk 1 above)
Comment (kmittal2) [reply fem/pgridfunc.cpp]: good catch @cjvogl

File: fem/pgridfunc.cpp (hunk 2 above)
Comment (cjvogl) [reply fem/pgridfunc.cpp]: resolved in https://github.com/mfem/mfem/pull/3588/commits/0bf3e290e043cdce8cbdb358250b044c87c7d02d, thanks!

**Review by cjvogl — CHANGES_REQUESTED** [review]
Remaining changes:
- proposed test
- address memory leak noted in https://github.com/mfem/mfem/pull/3588#discussion_r1223436484

File: fem/pgridfunc.cpp (hunk 1 above)
Comment (kmittal2) [reply fem/pgridfunc.cpp]: I fixed the unit test in latest commit by having the `GridFunction` own the `FiniteElementSpace` and `FECollection`.. this makes sure the `FiniteElementSpace` is deleted when the `GridFunction` is deleted..

File: tests/unit/fem/test_pgridfunc_save_serial.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+   Mesh x_ser_mesh = pmesh.GetSerialMesh(save_rank);
+   GridFunction x_par_to_ser = px.GetSerialGridFunction(save_rank, x_ser_mesh);
+   double par_to_ser_l2_err = 0.0;
+   if (my_rank == save_rank)
+   {
+      par_to_ser_l2_err = x_par_to_ser.ComputeL2Error(squaredFC);
+      REQUIRE(std::fabs(par_to_ser_l2_err-ser_l2_err) == MFEM_Approx(0.0));
+   }
+   //px.SaveAsSerial("paralleltoserial.gf");
```
Comment (cjvogl) [inline tests/unit/fem/test_pgridfunc_save_serial.cpp:77]: Perhaps I'm just missing something here, but with this line commented out, is the new `SaveAsSerial` tested?

Comment (kmittal2) [reply tests/unit/fem/test_pgridfunc_save_serial.cpp:77]: I will remove the commented out lines and rename the namespace from `save_in_serial` to `get_in_serial`. Does that sound good @cjvogl?

Comment (cjvogl) [reply tests/unit/fem/test_pgridfunc_save_serial.cpp:77]: I think it is indeed cleaner to not have commented out code without comments as to why the code is commented. that is not documented.  What is the harm in uncommenting the lines and obtaining full code coverage? Is it the creation of files left on disk?

Comment (kmittal2) [reply tests/unit/fem/test_pgridfunc_save_serial.cpp:77]: No harm really, the unit test as-is tests the new functionality.

Comment (cjvogl) [reply tests/unit/fem/test_pgridfunc_save_serial.cpp:77]: Perhaps we're talking past each other at this point... my concern is the unit test adds lines [947 through 959 ](https://github.com/mfem/mfem/pull/3588/files#diff-a1c11c4eae3e906589bdec42c41e2ed446e92bf76bb92e6f3ef35b5581875b8aR947-R959) that are **not** covered by the unit test as-is, i.e., if someone adds a bug to those lines, this unit test will not catch it as-is.

If there's no harm in uncommenting the line in question, then I think it should be uncommented to protect against such a scenario.  Or am I not understanding something that covers [those lines](https://github.com/mfem/mfem/pull/3588/files#diff-a1c11c4eae3e906589bdec42c41e2ed446e92bf76bb92e6f3ef35b5581875b8aR947-R959)

Comment (kmittal2) [reply tests/unit/fem/test_pgridfunc_save_serial.cpp:77]: Sounds good Chris.

Comment (kmittal2) [reply tests/unit/fem/test_pgridfunc_save_serial.cpp]: Done! Thanks for a thorough review @cjvogl.

Comment (cjvogl) [reply tests/unit/fem/test_pgridfunc_save_serial.cpp]: Thanks for the patience during the review!

**Comment by cjvogl** [comment]
The supplied unit test should be added to CMake build.  I've supplied a git patch that can be applied to do this.
[test_cmake.patch](https://github.com/mfem/mfem/files/11738141/test_cmake.patch)

File: tests/unit/fem/test_pgridfunc_save_serial.cpp (hunk 3 above)
Comment (kmittal2) [reply tests/unit/fem/test_pgridfunc_save_serial.cpp]: ah yes. my bad!

Comment (cjvogl) [reply tests/unit/fem/test_pgridfunc_save_serial.cpp]: Addressed in https://github.com/mfem/mfem/pull/3588/commits/e473491d06c7026cbd424b16576ba41672241107

**Review by cjvogl — APPROVED** [review]
Looks good to me!

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Pull request merged**